### PR TITLE
Added option to specify custom locations for SSL certificates

### PR DESCRIPTION
**Pull request by artisticcheese**

### CODE REVIEW

**Comment by pscripted** [comment]
First, much needed and appreciated clean up, thanks. Any reason you prefer the JSON hash for the cert location? I feel having them be proper (individual) class properties may be better.

**Comment by artisticcheese** [comment]
I think it might be easier for users to provide single property instead of 3 of them. If you really to prefer to have individual properties I can redo

**Comment by pscripted** [comment]
I understand the need, but thinking about it more, I think we may be covering up a bug by doing it this in a new property. This makes sense to be handled by the BaseConfigJSON property, where you can specify (in theory) anything this module doesn't cover. 

If you add the certs to that manually currently it will error out as it tries to add the properties separately when enabling TLS (including in your update), instead of reading it in from the baseconfigjson.  What do you think about something like this? Instead of a new property, specify your cert paths in the base config, and if they are there we consume them, if they are not we set them to a default.

config:
```
BaseConfigJson = @"
{
    "tlscacert":  "C:\\CustomCertPath\\myca.pem",
    "tlscert":  "C:\\CustomCertPath\\cert.pem",
    "tlskey":  "C:\\CustomCertPath\\key.pem"
}
"@
```
GetPendingConfiguration():
```
#Name of required docker config keys, and default file names
$tlskeys = @{"tlscacert"="ca.pem";"tlscert"="cert.pem";"tlskey"="key.pem"}
$fileExists = $true

if ($this.EnableTLS) {
    foreach ($cert in $tlskeys.GetEnumerator()) {
        if (-not $pendingConfiguration.psobject.Properties[$cert.name]) {          
            #Create the default if it's not there
            $pendingConfiguration | Add-Member -MemberType NoteProperty -Name $cert.name -Value "C:\ProgramData\docker\certs.d\$($cert.Value)"
        }
        if (-not (test-path $pendingConfiguration.psobject.Properties[$cert.name].Value )) { $fileExists =  $false }
    }
    if ($fileExists) {
      #If the certs exist, enable TLS
      $pendingConfiguration | Add-Member -MemberType NoteProperty -Name  "tlsverify" -Value $true   
    }
    else {
        foreach ($cert in $tlskeys.GetEnumerator()) {
            #without all the files, clear the key from the config to be written
            $pendingConfiguration.PSObject.Properties.Remove($cert.name)
        }
    }
}
else {
    foreach ($cert in $tlskeys.GetEnumerator()) {
        #if certs are specified in config but TLS is not enabled, remove them from the final config
        $pendingConfiguration.PSObject.Properties.Remove($cert.name)
    }
}
```

**Comment by artisticcheese** [comment]
Yes, this will work.

**Comment by artisticcheese** [comment]
So did you merge your changes? I want to add ability to have `drain` mode attribute to DSC to designate a node as a non worker node.

**Review by pscripted — APPROVED** [review]

**Comment by artisticcheese** [comment]
So your merged my changes instead of implementing yours as just passing objects directly via base config?

**Comment by pscripted** [comment]
I merged yours with all the other stuff you changed too, then added in the one section as above to use the base config. I'll do some testing of this branch before adding to master.

File: cDSCDockerSwarm.psm1 [hunk 1]
```diff
@@ -296,18 +283,24 @@ class cDockerConfig
     #Adds named pipe and TCP bindings to configuration, allowing external access. This is required for Swarm mode
     [DscProperty()]
     [boolean] $ExposeAPI
+
+    #Identify locations artifacts for TLS configuration
+    [DscProperty()]
```
Comment (artisticcheese) [inline cDSCDockerSwarm.psm1]: This property still exists? Would not you need to remove it?

**Comment by artisticcheese** [comment]
Waiting for completion so I can start working on adding "DrainMode" property

**Pull request merged**